File: Engine/src/Interval/interval.cpp
#include "interval.hpp"
#include <algorithm>

const Interval Interval::empty = Interval(infinity, -infinity);
const Interval Interval::universe = Interval(-infinity, infinity);

Interval::Interval() : min(+infinity), max(-infinity) {}

Interval::Interval(double min, double max) : min(min), max(max) {}

Interval::Interval(const Interval &a, const Interval &b) {
    min = a.min <= b.min ? a.min : b.min;
    max = a.max >= b.max ? a.max : b.max;
}

Interval Interval::expand(double delta) const {
    auto padding = delta / 2;
    return Interval(min - padding, max + padding);
}

double Interval::size() const { return max - min; }

bool Interval::contains(double x) const { return min <= x && x <= max; }

bool Interval::surrounds(double x) const { return min < x && x < max; }

double Interval::clamp(double x) const { return std::clamp(x, min, max); }

File: rust-version/src/interval.rs
use std::f64::{INFINITY, NEG_INFINITY};

#[derive(Debug, Copy, Clone)]
pub struct Interval {
    pub min: f64,
    pub max: f64,
}

#[allow(dead_code)]
impl Interval {
    pub const fn new(min: f64, max: f64) -> Self {
        Self { min, max }
    }

    pub const fn empty() -> Self {
        Self {
            min: INFINITY,
            max: NEG_INFINITY,
        }
    }

    pub const fn universe() -> Self {
        Self {
            min: NEG_INFINITY,
            max: INFINITY,
        }
    }

    pub fn size(&self) -> f64 {
        self.max - self.min
    }

    pub fn contains(&self, x: f64) -> bool {
        self.min <= x && x <= self.max
    }

    pub fn surrounds(&self, x: f64) -> bool {
        (self.min..self.max).contains(&x)
    }

    pub fn clamp(&self, x: f64) -> f64 {
        x.clamp(self.min, self.max)
    }
}
